command/job_stop.go

```diff
@@ -140,92 +140,141 @@ func (c *JobStopCommand) Run(args []string) int {
+			if err != nil {
+				c.Ui.Error(fmt.Sprintf("Error deregistering job with id %s err: %s", jobID, err))
+				statusCh <- 1
+				return
+			}
+
+			// TODO: does confirmation make sense with >1 jobs?
+			// getConfirmation := func(question string) (int, bool) {
+			// 	answer, err := c.Ui.Ask(question)
+			// 	if err != nil {
+			// 		c.Ui.Error(fmt.Sprintf("Failed to parse answer: %v", err))
+			// 		return 1, false
+			// 	}
+
+			// 	if answer == "" || strings.ToLower(answer)[0] == 'n' {
+			// 		// No case
+			// 		c.Ui.Output("Cancelling job stop")
+			// 		return 0, false
+			// 	} else if strings.ToLower(answer)[0] == 'y' && len(answer) > 1 {
+			// 		// Non exact match yes
+			// 		c.Ui.Output("For confirmation, an exact ‘y’ is required.")
+			// 		return 0, false
+			// 	} else if answer != "y" {
+			// 		c.Ui.Output("No confirmation detected. For confirmation, an exact 'y' is required.")
+			// 		return 1, false
+			// 	}
+			// 	return 0, true
+			// }
+
+			// // Confirm the stop if the job was a prefix match
+			// if jobID != *job.ID && !autoYes {
+			// 	question := fmt.Sprintf("Are you sure you want to stop job %q? [y/N]", *job.ID)
+			// 	code, confirmed := getConfirmation(question)
+			// 	if !confirmed {
+			// 		return code
+			// 	}
+			// }
+
+			// Confirm we want to stop only a single region of a multiregion job
+			// if job.IsMultiregion() && !global {
+			// 	question := fmt.Sprintf(
+			// 		"Are you sure you want to stop multi-region job %q in a single region? [y/N]", *job.ID)
+			// 	code, confirmed := getConfirmation(question)
+			// 	if !confirmed {
+			// 		return code
+			// 	}
+			// }
+
+			// Invoke the stop
+			opts := &api.DeregisterOptions{Purge: purge, Global: global, EvalPriority: evalPriority, NoShutdownDelay: noShutdownDelay}
+			wq := &api.WriteOptions{Namespace: jobs[0].JobSummary.Namespace}
+			evalID, _, err := client.Jobs().DeregisterOpts(*job.ID, opts, wq)
+			if err != nil {
+				c.Ui.Error(fmt.Sprintf("Error deregistering job with id %s err: %s", jobID, err))
+				statusCh <- 1
+				return
+			}
+
+			// If we are stopping a periodic job there won't be an evalID.
+			if evalID == "" {
+				statusCh <- 0
+				return
+			}
+
+			// Goroutine won't wait on monitor
+			if detach {
+				c.Ui.Output(evalID)
+				statusCh <- 0
+				return
+			}
+
+			// Start monitoring the stop eval
+			// and return result on status channel
+			mon := newMonitor(c.Ui, client, length)
+			statusCh <- mon.monitor(evalID)
+		}()
 	}
-	if len(jobs) > 1 {
-		if (jobID != jobs[0].ID) || (c.allNamespaces() && jobs[0].ID == jobs[1].ID) {
-			c.Ui.Error(fmt.Sprintf("Prefix matched multiple jobs\n\n%s", createStatusListOutput(jobs, c.allNamespaces())))
-			return 1
+	// users will still see
+	// errors if any while we
+	// wait for the goroutines
+	// to finish processing
+	wg.Wait()
+
+	// return a non-zero exit code
+	// if even a single job stop fails
+	for status := range statusCh {
```

Comment:
I don't think this will work since `range chan` iterates until the chan is closed, but nothing closes statusCh.

Iterating until `len(statusCh) == 0` is safe here since `wg.Wait()` above ensures all goroutines have exited.

Reply:
Thanks for pointing this out. That makes a lot of sense.

Alternatively, what do you think about `close(statusCh)` immediately after `wg.Wait()`. That should ensure no goroutine is going to write to the channel and we should be able to safely read the values out of it? That being said, It might not be completely idiomatic that the "receiver" here itself is closing the channel but this pattern seems safe. Let me know your thoughts?

Reply:
You're right, panic(ing) here seems like the right approach for this. Thanks for the clarification.

Reply:
I had one more thing I wanted to clear up, in order to test the changes for this PR, how do you suppose I go about it? Setting up Nomad and then doing a test run or is there another way using vagrant that I can find in the docs? If it's the latter, please ignore it, I'll go through it in a day.

Reply:
Thanks for that rundown on testing. I'll take a dig at the tests this weekend.